Add Order instance for NonEmptySet (#2779)

jatcwang opened the pull request:
- Add Serializable tests for NonEmptySet instances

Thank you for contributing to Cats!

This is a kind reminder to run `sbt prePR` and commit the changed files, if any, before submitting.

30sergey commented:
Дубликат #

30sergey commented:
#2783

kailuowang commented on core/src/main/scala/cats/data/NonEmptySet.scala:
```diff
@@ -417,3 +420,29 @@ sealed abstract private[data] class NonEmptySetInstances {
 }
+
+sealed abstract private[data] class NonEmptySetInstances0 extends NonEmptySetInstances1 {
+  implicit def catsDataOrderForNonEmptySetReal[A](implicit A: Order[A]): Order[NonEmptySet[A]] =
```
nit, is that `Real` suffix still needed?

jatcwang replied:
Not sure. catsDataOrderForNonEmptySet is the name of the existing one (which I removed the `implicit` keyword from). I can play around with it and see if it breaks bin compat

kailuowang replied:
did you mean `private[data] def catsDataEqForNonEmptySet[A: Order]:` ?

jatcwang replied:
yep since `NonEmptySetInstances` extends `NonEmptySetInstances0`

kailuowang replied:
:) `catsDataOrderForNonEmptySet` is different from `catsDataEqForNonEmptySet`.

jatcwang replied:
🤦‍♂️ well if MiMa is happy then I'm happy :)
I'm gonna try remove the old `catsDataEqForNonEmptySet` and see what happens

kailuowang replied:
sorry I wasn't clear. I meant that you didn't need to add a `Real` suffix, because it didn't conflict with anything else. It's the right approach to remove the `implict` keyword for `catsDataEqForNonEmptySet` , but you didn't need to add private[data] since the class is already `private[data]`

jatcwang replied:
yes I got you. What I was also going to do is to remove the original `catsDataEqForNonEmptySet`, since there's already a method named the same in `NonEmptySetInstances1`. (I think JVM's use of invokevirtual call means it won't break bin compat - so it'll be both source and binary compatible)

jatcwang replied:
Done

kailuowang commented on core/src/main/scala/cats/data/NonEmptySet.scala:
```diff
@@ -19,6 +19,8 @@ package data
 import cats.instances.sortedSet._
 import cats.kernel._
+import cats.syntax.order._
+import cats.instances.sortedSet._
```
duplicated from 2 lines above

jatcwang replied:
done

kailuowang commented:
I would like to propose :
We don't need two new classes, just one `NonEmptySetInstances0` is enough, you can place the `eq` instance there and the `order` instance in `NonEmptySetInstances`
I think this change is BC but Mima will raise a false alarm. We need to add a mima exception
here https://github.com/typelevel/cats/blob/master/build.sbt#L288
```
ProblemFilters.exclude[IncompatibleMethTypeProblem]("cats.data.NonEmptySetInstances.catsDataEqForNonEmptySet")
```

To make sure we didn't actually break BC, we need to add a line calling the instance def here https://github.com/typelevel/cats/blob/master/binCompatTest/src/main/scala/catsBC/MimaExceptions.scala similar to the other tests there.

jatcwang commented:
Done (Sorry I was AFK for a week). Simplified the instances hierachy as suggested and added a different test class for testing binary package private instances.

jatcwang commented:
CI is passing again @kailuowang :smile:

kailuowang commented on tests/src/test/scala/cats/tests/NonEmptySetSuite.scala:
```diff
@@ -20,17 +20,42 @@ package tests
+             SerializableTests.serializable(PartialOrder[NonEmptySet[ListWrapper[Int]]]))
+
+    Eq[NonEmptySet[ListWrapper[Int]]]
+    PartialOrder[NonEmptySet[ListWrapper[Int]]]
```
These two lines are unnecessary, no?

jatcwang replied:
These instance tests are written based on https://github.com/typelevel/cats/blob/4f92efedd2c30534c07b5f75a72b22e3e1ba9b13/tests/src/test/scala/cats/tests/NonEmptyListSuite.scala and from what I can gather is that it's to test that an `Eq` and `PartialOrder` instance can be resolved when an `Order` instance is in scope. 

Do you think they are redundant?

kailuowang replied:
The resolution is already tested in the law and serialization tests

jatcwang replied:
ok

kailuowang commented on tests/src/test/scala/cats/tests/NonEmptySetSuite.scala:
```diff
@@ -20,17 +20,42 @@ package tests
+    // NOTE: Order[NonEmptySet[A]] instance depends on Order[SortedSet[A]]. SortedSet only has Order instance which depends on Order[A] instance
+    //       and thus the only way we can get PartialOrder[NonEmptySet[A]] is via construction of Order[NonEmptySet[A]]
+    checkAll("NonEmptySet[ListWrapper[Int]]", PartialOrderTests[NonEmptySet[ListWrapper[Int]]].partialOrder)
+    checkAll("PartialOrder[NonEmptySet[ListWrapper[Int]]]",
```
Maybe I am missing something here, but this is covered by the `Order` test above, since we don't have a different way to construct a `PartialOrder`, not sure if we need this test?

jatcwang replied:
you're right. Removed thx

kailuowang commented on binCompatTest/src/test/scala/cats/data/CatsDataMimaException.scala:
```diff
@@ -0,0 +1,8 @@
+package cats.data
+import cats.implicits._
+
+object CatsDataPackagePrivateMimaExceptions {
```
My bad. I should've checked that mima no longer falsely claims that this type of change breaks BC. So this is no longer needed? Would you mine removing it?

jatcwang replied:
done
